### PR TITLE
Tls configuration

### PR DESCRIPTION
**Pull request by gezero**
## PR description

Adding two new options:
* `--rpc-http-tls-protocols` to be able to specify TLS protocol version 
* `--rpc-http-tls-ciphersuites` to be able to specify Java cypher suites

## Fixed Issue(s)

Fixes [#474](https://github.com/hyperledger/besu/issues/474): 

## Changelog

- [x] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 1]
```diff
@@ -1974,12 +1996,28 @@ private void checkPrivacyTlsOptionsDependencies() {
           "Known-clients file must be specified or CA clients must be enabled when TLS client authentication is enabled for JSON-RPC HTTP endpoint");
     }
 
+    rpcHttpTlsProtocols.retainAll(getJDKEnabledProtocols());
+    if (rpcHttpTlsProtocols.isEmpty()) {
+      throw new ParameterException(commandLine, "No valid TLS protocols specified");
```
Comment (fab-10) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:2001]: Could be useful to add the list of allowed protocols to the message

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 2]
```diff
@@ -1974,12 +1996,28 @@ private void checkPrivacyTlsOptionsDependencies() {
...
+    }
+
+    for (String ciphersuite : rpcHttpTlsCipherSuites) {
```
Comment (fab-10) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:2004]: why the validations of the cipherSuites is implemented differently from the tlsProtocol?

**Comment by fab-10** [comment]
You should add the `documentation` label to this PR and the related ticket, so that the doc team could update the Besu command line documentation

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 1 above)
Comment (gezero) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:2001]: Cool, I will add that

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 2 above)
Comment (gezero) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:2004]: There are only few protocols available (for instance on my jdk ["TLSv1.3", "TLSv1.2", "SSLv2Hello"]), their names are short and clear and you would only only diverge from default in legacy systems I would think. It makes sense to me to just show 1 error telling you that you did not specify any supported protocol I will update the error to show user which protocols are supported.


On the other hand when specifying the suite it requires precision so it makes sense to tell the user about each mistake explicitly. There are 48 enabled suites on my jdk. 

If you think I should remake it some other way, let me know 👍

Comment (fab-10) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:2004]: thanks for clarify, now it make sense the different approach

**Review by fab-10 — APPROVED** [review]
LGTM

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 3]
```diff
@@ -638,6 +642,20 @@ void setBannedNodeIds(final List<String> values) {
           "Enable to accept clients certificate signed by a valid CA for client authentication (default: ${DEFAULT-VALUE})")
   private final Boolean isRpcHttpTlsCAClientsEnabled = false;
 
+  @Option(
+      names = {"--rpc-http-tls-protocol", "--rpc-http-tls-protocols"},
+      description = "Comma separated list of TLS protocols to support (default: TLSv1.2)",
```
Comment (pinges) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:647]: TLS 1.3 should be (at least part of) the default. It gives better performance and is more secure and should be well supported by now. I guess first preference should be 1.3 and if we are worried about 1.3 support we could add 1.2.

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java [hunk 4]
```diff
@@ -2321,6 +2322,142 @@ public void rpcHttpTlsClientAuthWithCAClientAndKnownClientFile() {
...
+
+  @Test
+  public void rpcHttpTlsCheckValidProtocolsAndCipherSuites() {
+    final String host = "1.2.3.4";
+    final int port = 1234;
+    final String keystoreFile = "/tmp/test.p12";
+    final String keystorePasswordFile = "/tmp/test.txt";
+    final String protocols = "TLSv1.3,TLSv1.2";
+    final String cipherSuites =
+        "TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384,TLS_ECDHE_RSA_WITH_AES_128_GCM_SHA256";
```
Comment (pinges) [inline besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java:2423]: We should add a 1.3 cipher suite here, e.g. TLS_AES_256_GCM_SHA384

**Review by pinges — CHANGES_REQUESTED** [review]

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java (hunk 3 above)
Comment (gezero) [reply besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java]: You are right, updated.

File: besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java (hunk 4 above)
Comment (gezero) [reply besu/src/test/java/org/hyperledger/besu/cli/BesuCommandTest.java]: Updated

**Review by pinges — APPROVED** [review]
LGTM

**Pull request merged**